api.planx.uk/editor/copyFlow.ts

```diff
@@ -0,0 +1,42 @@
+import { Request, Response, NextFunction } from "express";
+import { makeUniqueFlow, getFlowData, insertFlow } from "../helpers";
+import { Flow } from "../types";
+
+const copyFlow = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+): Promise<Response | NextFunction | void> => {
+  try {
+    if (!req.user?.sub) {
+      return next({ status: 401, message: "User ID missing from JWT" });
+    }
+
+    // Fetch the original flow
+    const flow: Flow = await getFlowData(req.params.flowId);
+
+    // Ensure the newly copied flow has unique nodeIds
+    const randomReplacementCharacters = Math.random().toString(36).slice(2,6);
+    const uniqueFlowData = makeUniqueFlow(flow.data, randomReplacementCharacters);
```

Comment:
This is really worth thinking / talking about.

Nodes ids must be unique for flow navigation to work, we saw this issue with clones where non-unique node ids cause a lot of trouble.

From memory, the reason John was just replacing three characters was so that so could track if required where a flow came from (so there would be a relationship that devs could understand between `gtUKoDON` and `gtUKoLAM`). I don't think this has been something we've needed or used much, and especially if Editors can generate multiple copies this is going to have less and less value.

So my suggestion (please shoot holes in this!) would be -
 - Make the entire node Id unique to avoid issues with uniqueness in the long term
 - If we really care about understanding "where" a flow came from we can add a `flow_copy_events` audit table, or a `copied_from` column on flows. I'm not sure what the value of this would be apart from peace of mind?

Reply:
Yeah this totally works for me, it's a sensible approach here 👍

Reply:
This makes perfect sense to me, thanks for the summary 👌